Fix GRO830 Gazebo models issues... and QoL updates

MarcOlivierFecteau opened the pull request.

MarcOlivierFecteau committed:
Remove Gazebo warnings when loading models

This commit also applies auto-formatting.

CHANGES:

- Auto-format;
- Fix `Gazebo/<color>` material warnings;
- Comment out attributes that cause "undefined attribute in SDF model"
  warnings;
- Update Gazebo plugin names from Fortress to Harmonic (Jazzy compatible).

MarcOlivierFecteau committed:
Fix all world files for GRO830; Auto-format

This commit applies the changes from `4ff0088` and `11673d4` to all
relevant launch files in `racecar_gazebo/launch`. This should make it so
all lab exercises and the validation work without issues caused by the
models used by Gazebo. This commit also applies auto-formatting to the
changed files.

NOTES:

- This commit has NOT yet been tested on a faculty's lab PC. It is
  required to test the lab exercises AND the validation BEFORE merging
  this commit onto the main branch;
- Tests conducted and passed on my personal laptop (Ubuntu 24.04, ROS
  Jazzy).

lopetit reviewed (approved):
Ça semble ok de mon côté

MarcOlivierFecteau commented:
Confirmed working on lab PC with the following "tweaks" applied to the base image:

- `unset IGN_GAZEBO_RESOURCE_PATH` (shouldn't be set in the first place, because we use Gazebo Harmonic, which uses the `GZ_SIM_RESOURCE_PATH` environment variable)

The following "tweaks" were also applied to the build procedure:

- `colcon build --packages-skip serial pb2roscpp` (not needed for lab PCs, because there is no serial communication)

@lopetit, a note on these "tweaks" SHOULD be added to the educational material (in the lab's section), because the lab PCs' image can't be changed until august 2026.

MarcOlivierFecteau commented:
@MathieuMRodrigue, once you confirmed the changes are working on your side, could you take care of merging the PR? In my opinion, a "rebase and merge" is the most appropriate method.

chameau5050 commented on racecar_gazebo/launch/simulation.launch.py:
while we are there, can we make a general sim launch file and just include it in [racecar_tunnel_genie.launch.py](https://github.com/SherbyRobotics/racecar/pull/44/files#diff-0780b90ec3a88ceed8fb0bc65d89cfe77b19b6ba7aa182a0b44e58c47863b43b), [racecar_gazebo/launch/racecar_tunnel.launch.py](https://github.com/SherbyRobotics/racecar/pull/44/files#diff-2aabee06be08c2c8eb19d3854727206b5b563157af5be0d0f8ce6767df333752) [racecar_gazebo/launch/racecar_circuit.launch.py](https://github.com/SherbyRobotics/racecar/pull/44/files#diff-8bcbc236cc950563b02e8450a5c453fb1460310dbf48c29bc8ecc552ad695364) ? Because for now it is basically 3 time de same file

MarcOlivierFecteau replied:
Agreed. I didn't want to merge all files into one, because it would required modifications to the educational material, and since the class begins in a few weeks, I thought it would be better to leave the three launch files separate for A25, and merge them in an ulterior PR for A26.

MarcOlivierFecteau replied:
@chameau5050 @lopetit I merged the three files into one with the launch argument `world` to specify which world file is loaded into Gazebo (see [0d132b6](https://github.com/SherbyRobotics/racecar/pull/44/commits/0d132b6ce5999f427f8776af1173441b9bf7c157) for more details). This file implements the idea of the "general sim launch file" @chameau5050 mentionned (at least, my interpretation of it).

As I mentionned in my reply, this change involves significant modifications to the educational material, especially in the "lab" section, where all the Gazebo launch commands should be changed for `ros2 launch racecar_gazebo simulation.launch.py world:=<world_filename>`. Though, I do think the reduced number of LoC and files and the fact that there is only one command to start the Gazebo simulation do make this change worth it.

@lopetit Since I don't have access to the student guide's source file, I let you take care of the modifications (unless you share it to me; then I'd gladly take the responsibility).

chameau5050 replied:
@MarcOlivierFecteau that is mostly what I had in mind. To not change the student guide I is also possible the keep the old launch file but to simplify them so they just contante the include of ros2 launch racecar_gazebo simulation.launch.py world:=<world_filename>. But your new implemetation is more clean and minimalist.

chameau5050 reviewed (commented):
maybe a tiny change while we are there

lopetit commented:
@MarcOlivierFecteau can you add these 2 tweaks to [doc/gro830-lab-a25.md](https://github.com/SherbyRobotics/racecar/blob/ros2/doc/gro830-lab-a25.md)? I have already added a link to this in the educational material.

MarcOlivierFecteau committed:
Remove linter warning for `robot_description_config`'s type

Additionally, this commit removes trailing spaces.

NOTES:

- This is a "QoL" commit: no tests are required.

MarcOlivierFecteau committed:
Merge all "start simulation" launch files into one

The three launch files used to start the Gazebo simulation were
basically the same, with only the `.world` file name being different.
Ergo, I took the identical code from the launch files, and added an
argument to specify the world file loaded into Gazebo.

NOTES:

- This change involves non-negligible modifications to the documentation
  in the educational material;
- This commit has NOT yet been tested on a faculty's lab PC, and MUST be
  tested BEFORE merging the PR into the main branch.

TODO:

- Test the "unity" launch file on the faculty's lab PCs (ideally, by
  several people).

MarcOlivierFecteau committed:
Add documentation for tweaks for the faculty's lab PCs

MarcOlivierFecteau commented:
> @MarcOlivierFecteau can you add these 2 tweaks to [doc/gro830-lab-a25.md](https://github.com/SherbyRobotics/racecar/blob/ros2/doc/gro830-lab-a25.md)? I have already added a link to this in the educational material.

Documentation for both tweaks has been added to [doc/gro830-lab-a25.md](https://github.com/SherbyRobotics/racecar/blob/ros2/doc/gro830-lab-a25.md).

MarcOlivierFecteau committed:
Fix string interpolation issues introduced in 0d132b6

This makes the "general launch file" for starting the Gazebo simulation
work.

NOTES:

- Do NOT use f-strings inside `IncludeLaunchDescription`! As a general
  rule, try to avoid f-strings in ROS2 lanch files: use the
  `launch.substitutions` module instead;
- This commit has been tested on my personal laptop (Ubuntu 24.04, ROS
  Jazzy, Gazebo Harmonic).

TODO:

- Test this commit on a faculty's lab PC (ideally, with both
  `IGN_GAZEBO_RESOURCE_PATH` and `GZ_SIM_RESOURCE_PATH` used as
  environment variable, to take into account potential oversights from
  students... and tutors / technical support staff).

chameau5050 commented on racecar_gazebo/launch/simulation.launch.py:
```diff
@@ -0,0 +1,66 @@
+                    [
+                        racecar_gazebo,
+                        "worlds",
+                        ["racecar_", LaunchConfiguration("world"), ".world"],
```
I do prefer the method you were using before (with path.join) since in allow more complex behavior. I believe the bug you were having with the previous version was that you were not using an opaque function to extract the context of the world argument [see](https://github.com/SherbyRobotics/racecar/blob/11673d43207e911cd702fe439f248bab35ea0181/racecar_navigation/launch/slam.launch.py#L17) for an example of what I mean

MarcOlivierFecteau replied:
@chameau5050 [c104e86](https://github.com/SherbyRobotics/racecar/pull/44/commits/c1604e85a9b08ade312f6e82e196fa9c28bbda74) implements the request changes, and is confirmed working on a faculty's lab PC using the procedure described [here](https://github.com/SherbyRobotics/racecar/pull/44#issuecomment-3442876990).

chameau5050 reviewed (changes requested):
good job, but I do prefer the old method

MarcOlivierFecteau commented:
[e59279f](https://github.com/SherbyRobotics/racecar/pull/44/commits/e59279f6461f49c918ca342c3cabcb69d45ba424) confirmed working on a lab PC with both `IGN_GAZEBO_RESOURCE_PATH` being set and unset (because the launch file creates `GZ_SIM_RESOURCE_PATH`, I believe Gazebo uses this environment variable and doesn't fallback on the deprecated environment variable). Below are the commands added in `.bashrc`:

```bash
source /opt/ros/jazzy/setup.bash
if [ -f ~/ros2_ws/install/local_setup.bash ]; then
  source ~/ros2_ws/install/local_setup.bash
fi
unset IGN_GAZEBO_RESOURCE_PATH  # Also tested with this line commented
```

Below are the commands used for the test:

```bash
cd ~/ros2_ws
colcon build --packages-skip serial pb2roscpp
source install/local_setup.bash
ros2 launch racecar_gazebo simulation.launch.py  # Test 1: default value
ros2 launch racecar_gazebo simulation.launch.py world:=tunnel  # Test 2: lab world (navigation)
ros2 launch racecar_gazebo simulation.launch.py world:=tunnel_genie  # Test 3: validation
ros2 launch racecar_gazebo simulation.launch.py world:=circuit  # Test 4: lab world (behaviors)
```

MarcOlivierFecteau committed:
Use `OpaqueFunction` to correctly parse `world` launch argument

This commit implements [this requested change](https://github.com/SherbyRobotics/racecar/pull/44/files/e59279f6461f49c918ca342c3cabcb69d45ba424#r2459951687).
It essentially "rolls back" to using Python f-strings for string
interpolation of the path to the world file loaded in Gazebo instead of
using ROS's `launch.substitutions` module.

TODO:

- Test this commit on a faculty's lab PC BEFORE merging the PR.

chameau5050 reviewed (approved):
Looks good to me ! Good job.

MarcOlivierFecteau committed:
Fix SLAM errors introduced in ab34974

The changes in ab34974 aimed to remove warnings when starting the Gazebo
simulation caused by a discrepancy in attributes between the `.gazebo`
file and the SDF model. However, commenting out these attributes made
the SLAM stopped working because it couldn't properly convert the LiDAR
messages, which in turn caused it so the map couldn't be fetched.

This commit rolls back the changes, and the SLAM now works properly.

NOTES:

- The attributes warnings have reappeared. Eventually, we will need to
address this issue, but for now it is considered low priority.
